File: BE/internal/service/platform_discount_service.go
package service

import (
	"context"
	"errors"
	"time"

	"github.com/google/uuid"

	"rrnet/internal/domain/billing"
	"rrnet/internal/repository"
)

var (
	ErrPlatformDiscountCodeRequired = errors.New("discount code is required")
	ErrPlatformDiscountCodeTaken    = errors.New("discount code already exists")
)

type PlatformDiscountService struct {
	repo *repository.PlatformDiscountRepository
}

func NewPlatformDiscountService(repo *repository.PlatformDiscountRepository) *PlatformDiscountService {
	return &PlatformDiscountService{repo: repo}
}

type CreatePlatformDiscountRequest struct {
	Code        string                       `json:"code"`
	Name        string                       `json:"name"`
	Description *string                      `json:"description,omitempty"`
	Type        billing.PlatformDiscountType `json:"type"`
	Value       float64                      `json:"value"`
	MinPurchase float64                      `json:"min_purchase"`
	MaxDiscount *float64                     `json:"max_discount,omitempty"`
	UsageLimit  *int                         `json:"usage_limit,omitempty"`
	ExpiresAt   *time.Time                   `json:"expires_at,omitempty"`
	IsActive    bool                         `json:"is_active"`
}

type UpdatePlatformDiscountRequest = CreatePlatformDiscountRequest

func (s *PlatformDiscountService) Create(ctx context.Context, req *CreatePlatformDiscountRequest) (*billing.PlatformDiscount, error) {
	if req.Code == "" {
		return nil, ErrPlatformDiscountCodeRequired
	}

	exists, err := s.repo.CodeExists(ctx, req.Code, nil)
	if err != nil {
		return nil, err
	}
	if exists {
		return nil, ErrPlatformDiscountCodeTaken
	}

	now := time.Now()
	d := &billing.PlatformDiscount{
		ID:          uuid.New(),
		Code:        req.Code,
		Name:        req.Name,
		Description: req.Description,
		Type:        req.Type,
		Value:       req.Value,
		MinPurchase: req.MinPurchase,
		MaxDiscount: req.MaxDiscount,
		UsageLimit:  req.UsageLimit,
		UsedCount:   0,
		ExpiresAt:   req.ExpiresAt,
		IsActive:    req.IsActive,
		CreatedAt:   now,
		UpdatedAt:   now,
	}

	if err := s.repo.Create(ctx, d); err != nil {
		return nil, err
	}
	return d, nil
}

func (s *PlatformDiscountService) GetByID(ctx context.Context, id uuid.UUID) (*billing.PlatformDiscount, error) {
	return s.repo.GetByID(ctx, id)
}

func (s *PlatformDiscountService) GetByCode(ctx context.Context, code string) (*billing.PlatformDiscount, error) {
	return s.repo.GetByCode(ctx, code)
}

func (s *PlatformDiscountService) List(ctx context.Context, includeInactive bool) ([]*billing.PlatformDiscount, error) {
	return s.repo.List(ctx, includeInactive)
}

func (s *PlatformDiscountService) Update(ctx context.Context, id uuid.UUID, req *UpdatePlatformDiscountRequest) (*billing.PlatformDiscount, error) {
	d, err := s.repo.GetByID(ctx, id)
	if err != nil {
		return nil, err
	}

	exists, err := s.repo.CodeExists(ctx, req.Code, &id)
	if err != nil {
		return nil, err
	}
	if exists {
		return nil, ErrPlatformDiscountCodeTaken
	}

	d.Code = req.Code
	d.Name = req.Name
	d.Description = req.Description
	d.Type = req.Type
	d.Value = req.Value
	d.MinPurchase = req.MinPurchase
	d.MaxDiscount = req.MaxDiscount
	d.UsageLimit = req.UsageLimit
	d.ExpiresAt = req.ExpiresAt
	d.IsActive = req.IsActive
	d.UpdatedAt = time.Now()

	if err := s.repo.Update(ctx, d); err != nil {
		return nil, err
	}
	return d, nil
}

func (s *PlatformDiscountService) Delete(ctx context.Context, id uuid.UUID) error {
	return s.repo.Delete(ctx, id)
}

func (s *PlatformDiscountService) ValidateCode(ctx context.Context, code string, amount float64) (*billing.PlatformDiscount, float64, error) {
	d, err := s.repo.GetByCode(ctx, code)
	if err != nil {
		return nil, 0, err
	}

	if !d.IsValid(amount) {
		return nil, 0, errors.New("discount code is not applicable or has expired")
	}

	discountAmount := d.CalculateDiscount(amount)
	return d, discountAmount, nil
}

File: fe/src/lib/api/platformDiscountService.ts
import apiClient from "./apiClient";

export type PlatformDiscountType = "percent" | "nominal";

export interface PlatformDiscount {
  id: string;
  code: string;
  name: string;
  description?: string;
  type: PlatformDiscountType;
  value: number;
  min_purchase: number;
  max_discount?: number;
  usage_limit?: number;
  used_count: number;
  expires_at?: string;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

export interface CreatePlatformDiscountRequest {
  code: string;
  name: string;
  description?: string;
  type: PlatformDiscountType;
  value: number;
  min_purchase: number;
  max_discount?: number;
  usage_limit?: number;
  expires_at?: string;
  is_active: boolean;
}

export const platformDiscountService = {
  list: async (includeInactive = true): Promise<PlatformDiscount[]> => {
    const response = await apiClient.get<{ data: PlatformDiscount[]; total: number }>(
      `/superadmin/billing/discounts?include_inactive=${includeInactive}`
    );
    return response.data.data || [];
  },

  get: async (id: string): Promise<PlatformDiscount> => {
    const response = await apiClient.get(`/superadmin/billing/discounts/${id}`);
    return response.data;
  },

  create: async (data: CreatePlatformDiscountRequest): Promise<PlatformDiscount> => {
    const response = await apiClient.post("/superadmin/billing/discounts", data);
    return response.data;
  },

  update: async (id: string, data: CreatePlatformDiscountRequest): Promise<PlatformDiscount> => {
    const response = await apiClient.put(`/superadmin/billing/discounts/${id}`, data);
    return response.data;
  },

  delete: async (id: string): Promise<void> => {
    await apiClient.delete(`/superadmin/billing/discounts/${id}`);
  },

  validate: async (code: string, amount: number) => {
    const response = await apiClient.post("/public/validate-discount", { code, amount });
    return response.data;
  },

  apply: async (invoice_id: string, code: string) => {
    const response = await apiClient.post("/public/apply-discount", { invoice_id, code });
    return response.data;
  },

  remove: async (invoice_id: string) => {
    const response = await apiClient.post("/public/remove-discount", { invoice_id });
    return response.data;
  },
};
